grails-app/controllers/au/org/ala/collectory/TokenInterceptor.groovy

```diff
@@ -0,0 +1,133 @@
+package au.org.ala.collectory
+
+import au.org.ala.ws.security.JwtProperties
+import groovy.transform.CompileStatic
+import groovy.util.logging.Slf4j
+import org.apache.commons.lang3.tuple.Pair
+import org.pac4j.core.adapter.FrameworkAdapter
+import org.pac4j.core.client.DirectClient
+import org.pac4j.core.config.Config
+import org.pac4j.core.context.CallContext
+import org.pac4j.core.context.WebContext
+import org.pac4j.core.context.session.SessionStore
+import org.pac4j.core.credentials.Credentials
+import org.pac4j.core.exception.CredentialsException
+import org.pac4j.core.profile.ProfileManager
+import org.pac4j.core.profile.UserProfile
+import org.pac4j.jee.context.JEEFrameworkParameters
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.beans.factory.annotation.Qualifier
+import org.springframework.boot.context.properties.EnableConfigurationProperties
+import javax.annotation.PostConstruct
+
+@CompileStatic
+@Slf4j
+@EnableConfigurationProperties(JwtProperties)
+/**
+ * This interceptor ALWAYS VALIDATE JWT token and fill scopes to the user profile in request.
+ * It can be removed when the new security plugin >7.0.1 is used, as it will handle the token validation and profile management.
+ */
+class TokenInterceptor {
```

Comment:
When testing with user JWT this interceptor causes a later failure because the `roles` are not included in the user profile. Add this functionality.

Reply:
I need a detailed example for debugging